src/main/java/io/confluent/idesidecar/restapi/messageviewer/RecordDeserializer.java

```diff
@@ -0,0 +1,324 @@
+      String topic,
+      boolean isKey
+  ) {
+    return deserialize(bytes, schemaRegistryClient, topic, isKey, Optional.empty());
+  }
+
+  private static void cacheSchemaFetchError(RestClientException e, int schemaId) {
+    var retryTime = Instant.now().plus(CACHE_FAILED_SCHEMA_ID_FETCH_DURATION);
+    var timeFormatter = DateTimeFormatter.ofPattern("HH:mm:ss")
+        .withZone(ZoneId.systemDefault());
+    Log.errorf(
+        "Failed to retrieve schema with ID %d. Will try again in %d seconds at %s. Error: %s",
+        schemaId,
+        CACHE_FAILED_SCHEMA_ID_FETCH_DURATION.getSeconds(),
+        timeFormatter.format(retryTime),
+        e.getMessage(),
+        e
+    );
+    var errorMessage = String.format(
+        "Failed to retrieve schema with ID %d: %s",
+        schemaId,
+        e.getMessage()
+    );
+    SCHEMA_FETCH_ERROR_CACHE.put(schemaId, errorMessage);
+  }
+
+  /**
+   * Handles the trivial cases of deserialization: null, empty, or non-Schema Registry encoded data.
+   *
+   * @param bytes                The byte array to parse
+   * @param schemaRegistryClient The SchemaRegistryClient used for deserialization of
+   *                             Schema Registry encoded data
+   * @return An Optional containing a DecodedResult if the base cases are met, or empty otherwise
+   */
+  private static Optional<DecodedResult> maybeTrivialCase(
+      byte[] bytes,
+      SchemaRegistryClient schemaRegistryClient
+  ) {
+    // If the byte array is null or empty, we return a NullNode or an empty string, respectively.
+    if (bytes == null) {
+      return Optional.of(new DecodedResult(NullNode.getInstance(), null));
+    }
+    if (bytes.length == 0) {
+      return Optional.of(new DecodedResult(new TextNode(""), null));
+    }
+
+    // If the first byte is not the magic byte, we try to parse the data as a JSON object
+    // or fall back to a string if parsing fails. Simple enough.
+    if (bytes[0] != MAGIC_BYTE) {
+      return Optional.of(new DecodedResult(safeReadTree(bytes), null));
+    }
+
+    // If the first byte is the magic byte, but we weren't provided a schema registry client,
+    // we can't decode the data, so we just return the raw bytes.
+    if (schemaRegistryClient == null) {
+      return Optional.of(new DecodedResult(
+          safeReadTree(bytes),
+          "The value references a schema but we can't find the schema registry"
+      ));
+    }
+
+    // Alright, let's move on to more complex cases.
+    return Optional.empty();
+  }
+
+  /**
+   * Try to read the byte array as a JSON object, falling back to a string if it fails.
+   * @param bytes The byte array to read
+   * @return      A JsonNode representing the byte array as a JSON object, or a TextNode
+   */
+  private static JsonNode safeReadTree(byte[] bytes) {
+    try {
+      return OBJECT_MAPPER.readTree(bytes);
+    } catch (IOException e) {
+      return TextNode.valueOf(new String(bytes, StandardCharsets.UTF_8));
+    }
+  }
```

Comment:
I think you need to check if the result of `OBJECT_MAPPER.readTree` [is a JSON object](https://fasterxml.github.io/jackson-databind/javadoc/2.8/com/fasterxml/jackson/databind/JsonNode.html#isObject()) because it can also return other node types:

```java
OBJECT_MAPPER
  .readTree("12345".getBytes(StandardCharsets.UTF_8))
  .getClass()
  .getCanonicalName();
// Returns a com.fasterxml.jackson.databind.node.IntNode

OBJECT_MAPPER
  .readTree("\"12345\"".getBytes(StandardCharsets.UTF_8))
  .getClass()
  .getCanonicalName();
// Returns a com.fasterxml.jackson.databind.node.TextNode

OBJECT_MAPPER
  .readTree("true".getBytes(StandardCharsets.UTF_8))
  .getClass()
  .getCanonicalName();
// Returns a com.fasterxml.jackson.databind.node.BooleanNode

OBJECT_MAPPER
  .readTree("false".getBytes(StandardCharsets.UTF_8))
  .getClass()
  .getCanonicalName();
// Returns a com.fasterxml.jackson.databind.node.BooleanNode

OBJECT_MAPPER
  .readTree("null".getBytes(StandardCharsets.UTF_8))
  .getClass()
  .getCanonicalName();
// Returns a com.fasterxml.jackson.databind.node.NullNode

OBJECT_MAPPER
  .readTree("{}".getBytes(StandardCharsets.UTF_8))
  .getClass()
  .getCanonicalName();
// Returns a com.fasterxml.jackson.databind.node.ObjectNode

OBJECT_MAPPER
  .readTree("{\"key\": \"12345\"}".getBytes(StandardCharsets.UTF_8))
  .getClass()
  .getCanonicalName();
// Returns a com.fasterxml.jackson.databind.node.ObjectNode

OBJECT_MAPPER
  .readTree("{\"key\": \"12345\"".getBytes(StandardCharsets.UTF_8))
  .getClass()
  .getCanonicalName();
// Throws a com.fasterxml.jackson.core.io.JsonEOFException with the message "Unexpected end-of-input: expected close marker for Object (start marker at [Source: REDACTED (`StreamReadFeature.INCLUDE_SOURCE_IN_LOCATION` disabled); line: 1, column: 1]) at [Source: REDACTED (`StreamReadFeature.INCLUDE_SOURCE_IN_LOCATION` disabled); line: 1, column: 16]"
// Would fall back to TextNode, which is great
```

Reply:
Those other node types are also valid `JsonNode`s I believe, which is the behavior what we want. 

The intention here is to maybe convert a stream of bytes to valid JSON - number, string, list, object, boolean, and if that's not possible, we return the UTF-8 representation of the data as a `TextNode` like you noted.

Reply:
> Those other node types are also valid JsonNodes I believe, which is the behavior we want.
> The intention here is to maybe convert a stream of bytes to valid JSON - number, string, list, object, boolean, and if that's not possible, we return the UTF-8 representation of the data as a TextNode like you noted.

Yep - they're valid `JsonNode`s but not JSON objects as suggested by the JavaDoc.

So far, the sidecar's consume API returned a string or a JSON object (if a schema is used) for record keys/values. The current implementation in your PR would change the behavior and, for instance, return a JSON number if it can be parsed from the byte array. I'm not sure if we want to change this.

Reply:
Tested this change with the extension message viewer and so far I've not encountered any issues. Even the search functionality continues to work just fine.

Reply:
Thanks for checking it!